src/txmempool.h

```diff
@@ -619,8 +619,10 @@ class CTxMemPool
     /** Expire all transaction (and their dependencies) in the mempool older than time. Return the number of removed transactions. */
     int Expire(int64_t time);
 
-    /** Returns false if the transaction is in the mempool and not within the chain limit specified. */
-    bool TransactionWithinChainLimit(const uint256& txid, size_t chainLimit) const;
+    /** Get the maximum of the chain limit value in the mempool of the given transaction.
+      * If the transaction is not in the mempool, -1 is returned.
```

Comment:
if not in the mempool could return 0 since `GetCountWithAncestors` and `GetCountWithDescendants` are always `> 0` (self is counted). In that case you can revert s/uint64_t/int64_t.

Reply:
I looked in the code and it seems to me they can be both 0 and that self is not counted.

Reply:
@NicolasDorier have you checked these?
https://github.com/bitcoin/bitcoin/blob/a34ac6ae0788b50e12dd2c8ac59dbda2a03a3c2e/src/txmempool.cpp#L30
https://github.com/bitcoin/bitcoin/blob/a34ac6ae0788b50e12dd2c8ac59dbda2a03a3c2e/src/txmempool.cpp#L36
https://github.com/bitcoin/bitcoin/blob/a34ac6ae0788b50e12dd2c8ac59dbda2a03a3c2e/src/txmempool.cpp#L317
https://github.com/bitcoin/bitcoin/blob/a34ac6ae0788b50e12dd2c8ac59dbda2a03a3c2e/src/txmempool.cpp#L326

Reply:
woops, completely missed that indeed. I failed to notice how the clients were forcing this invariant though.

Reply:
Got confused when I saw [that](https://github.com/bitcoin/bitcoin/blob/9501dc27b336fc20adb0c367e0904b57bd507f51/src/txmempool.cpp#L857) which seemed to me it could be 0.

Reply:
I made the change for a reason but that reason may have been flawed. I'll try it with uint64_t and 0 again. Thanks!

Reply:
The thing is, the check in `SelectCoinsMinConf` checks if the chain limit value `>= nMaxAncestors`, which is `0` for `no chained ancestors` in the code. With the old method, this would be true, but with a 0 returned, it would be false.

I could change the `0` to a `1` to fix it, which may be a better solution here, and is probably what I will end up doing.

Reply:
If I do that, the semantics becomes misleading. It says 'max *ancestors*' so 1 makes it sound like it allows 1 ancestor or descendant, when in reality it only counts itself. I'm gonna keep the change to signed, for now.

Reply:
Indeed, now:

with @promag suggestion,  if `nMaxAncestors==0` and the tx is not in mempool then `mempool.ChainLimitValue(pcoin->GetHash()) >= nMaxAncestors` evaluate to `true`.

While before `!mempool.TransactionWithinChainLimit(pcoin->GetHash(), nMaxAncestors)` evaluate to `false`.